### ISSUE TITLE
test_iocp fails on Python 3 under Windows, test_tcp hangs on Python 3 under Windows

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8631]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)| @rodrigc reported|
|-|-|
|Trac ID|trac#8631|
|Type|enhancement|
|Created|2016-07-20 03:25:51Z|
|Branch|https://github.com/twisted/twisted/tree/8631-rodrigc-iocp-py3|

While trying to test Twisted under Windows 10, with Python 3.5, I found that:

```
trial twisted.internet.test.test_iocp
```

would fail

```
trial twisted.internet.test.test_tcp
```

would hang.

Under Python 2.7, everything works.

<details><summary>Searchable metadata</summary>

```
trac-id__8631 8631
type__enhancement enhancement
reporter__rodrigc rodrigc
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8631_rodrigc_iocp_py3 8631-rodrigc-iocp-py3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1468985151642831 1468985151642831
changetime__1471039832007165 1471039832007165
version__None None
owner__None None

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_2"></a>|@rodrigc commented|
|-|-|

See:
http://twistedmatrix.com/trac/ticket/8631

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

I assume that was supposed to be a link to https://github.com/twisted/twisted/pull/359 and not back to this ticket :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @rodrigc|
|-|-|

Thanks for the fix.  It looks like just one windows builder (`windows7-64-py2.7`) is having some issues with this code.  If any updates are needed to the buildbot to fix it so this builds properly just let me know; I am a bit confused why it's working on the many other builders.

It generally looks like this includes some good fixes, migrates us to non-deprecated APIs, and might fix a few memory leaks.  However, I am not sure all of it is correct, so I left some comments on the github PR about a couple of doc coverage issues and encoding concerns.  Can you address those and resubmit?  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_5"></a>|@rodrigc commented|
|-|-|

Patch updated

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)|@rodrigc removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Approved & merged.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set status to `closed`|
|-|-|

Oh, ugh. So excited about the passing build I mis-formatted the commit message.  It was fixed in 9bc4336.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_9"></a>|@rodrigc commented|
|-|-|

Fixed in [9bc4336478 ](https://twistedmatrix.com/trac/changeset/9bc4336478cd1ed4ee25911cd81b8b5a4de3ad54)
